### ISSUE TITLE
Re: [uClibc] perl-5.8.5: non-portable code causes build failure on uclibc

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7566]
Migrated from [rt.perl.org#32197](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=32197) (status was 'resolved')

Searchable as RT32197$

**Comment by p5pRT** [comment]
### From vda@port.imtp.ilyichevsk.odessa.ua

This part of perl source code caused build failure on uclibc\.

perl\-5\.8\.5/perlio\.c&#8203;:

static int
PerlIOStdio\_invalidate\_fileno\(pTHX\_ FILE \*f\)
\{
&nbsp;    /\* XXX this could use PerlIO\_canset\_fileno\(\) and
&nbsp;     \* PerlIO\_set\_fileno\(\) support from Configure
&nbsp;     \*/
\#  if defined\(\_\_GLIBC\_\_\)
&nbsp;    /\* There may be a better way for GLIBC&#8203;:
&nbsp;        \- libio\.h defines a flag to not close\(\) on cleanup
&nbsp;     \*/
&nbsp;    f\->\_fileno = \-1;
&nbsp;    return 1;
\#  elif defined\(\_\_sun\_\_\)
\.\.\.\.


This is a gross hack\. Why do perl fiddle with C lib internals
which are meant to be opaque to the user?

It is used exactly in one place only&#8203;:


IV
PerlIOStdio\_close\(pTHX\_ PerlIO \*f\)
\{
&nbsp;    FILE \*stdio = PerlIOSelf\(f\, PerlIOStdio\)\->stdio;
&nbsp;    if \(\!stdio\) \{
&nbsp;    errno = EBADF;
&nbsp;    return \-1;
&nbsp;    \}
&nbsp;    else \{
&nbsp;        \.\.\.\.
&nbsp;    if \(invalidate\) \{
&nbsp;           /\* For STD\* handles don't close the stdio at all
&nbsp;           this is because we have shared the FILE \* too
&nbsp;            \*/
&nbsp;        if \(stdio == stdin\) \{
&nbsp;            /\* Some stdios are buggy fflush\-ing inputs \*/
&nbsp;            return 0;
&nbsp;        \}
&nbsp;        else if \(stdio == stdout || stdio == stderr\) \{
&nbsp;            return PerlIO\_flush\(f\);
&nbsp;        \}
&nbsp;            /\* Tricky \- must fclose\(stdio\) to free memory but not close\(fd\)
&nbsp;           Use Sarathy's trick from maint\-5\.6 to invalidate the
&nbsp;           fileno slot of the FILE \*
&nbsp;        \*/
&nbsp;        result = PerlIO\_flush\(f\);
&nbsp;        saveerr = errno;
&nbsp;            if \(\!\(invalidate = PerlIOStdio\_invalidate\_fileno\(aTHX\_ stdio\)\)\) \{
&nbsp;            dupfd = PerlLIO\_dup\(fd\);
&nbsp;        \}
&nbsp;    \}
&nbsp;        \.\.\.\.\.\.
\-\-
vda

**Comment by p5pRT** [comment]
### From jcromie@divsol.com

Denis Vlasenko \(via RT\) wrote&#8203;:

>\# New Ticket Created by  Denis Vlasenko 
>\# Please include the string&#8203;:  \[perl \#32189\]
>\# in the subject line of all future correspondence about this issue\. 
>\# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=32189 >
>
>
>This part of perl source code caused build failure on uclibc\.
>
>perl\-5\.8\.5/perlio\.c&#8203;:
>
>static int
>PerlIOStdio\_invalidate\_fileno\(pTHX\_ FILE \*f\)
>\{
>    /\* XXX this could use PerlIO\_canset\_fileno\(\) and
>     \* PerlIO\_set\_fileno\(\) support from Configure
>     \*/
>\#  if defined\(\_\_GLIBC\_\_\)
>    /\* There may be a better way for GLIBC&#8203;:
>        \- libio\.h defines a flag to not close\(\) on cleanup
>     \*/
>    f\->\_fileno = \-1;
>    return 1;
>\#  elif defined\(\_\_sun\_\_\)
>\.\.\.\.
>
>
>This is a gross hack\. Why do perl fiddle with C lib internals
>which are meant to be opaque to the user?
>
>  
>
for speed\.  You can disable it with a configure adjustment\.
Heres what Ive got for 5\.8\.5 on linux\. 

$ perl \-V&#8203;:\.\\\*fast\.\\\*
d\_faststdio='undef';
usefaststdio='define';

Note that it was undef in config\, and apparently overridden \(its not a 
hint\-file entry in linux\)

btw \- ICBW\.  If so\, im sure someone will correct me\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From mjn3@codepoet.org

On Thu\, Oct 28\, 2004 at 10&#8203;:42&#8203;:49AM \+0300\, Denis Vlasenko wrote&#8203;:
> This part of perl source code caused build failure on uclibc\.
> 
> perl\-5\.8\.5/perlio\.c&#8203;:

See http&#8203;://www\.uclibc\.org/lists/uclibc/2004\-March/008512\.html

Manuel

**Comment by p5pRT** [comment]
### From vda@port.imtp.ilyichevsk.odessa.ua

On Thursday 28 October 2004 13&#8203;:14\, Jim Cromie via RT wrote&#8203;:
> Denis Vlasenko \(via RT\) wrote&#8203;:
> 
> >\# New Ticket Created by  Denis Vlasenko 
> >\# Please include the string&#8203;:  \[perl \#32189\]
> >\# in the subject line of all future correspondence about this issue\. 
> >\# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=32189 >
> >
> >
> >This part of perl source code caused build failure on uclibc\.
> >
> >perl\-5\.8\.5/perlio\.c&#8203;:
> >
> >static int
> >PerlIOStdio\_invalidate\_fileno\(pTHX\_ FILE \*f\)
> >\{
> >    /\* XXX this could use PerlIO\_canset\_fileno\(\) and
> >     \* PerlIO\_set\_fileno\(\) support from Configure
> >     \*/
> >\#  if defined\(\_\_GLIBC\_\_\)
> >    /\* There may be a better way for GLIBC&#8203;:
> >        \- libio\.h defines a flag to not close\(\) on cleanup
> >     \*/
> >    f\->\_fileno = \-1;
> >    return 1;
> >\#  elif defined\(\_\_sun\_\_\)
> >\.\.\.\.
> >
> >
> >This is a gross hack\. Why do perl fiddle with C lib internals
> >which are meant to be opaque to the user?
> >
> >  
> >
> for speed\.  You can disable it with a configure adjustment\.
> Heres what Ive got for 5\.8\.5 on linux\. 
> 
> $ perl \-V&#8203;:\.\\\*fast\.\\\*
> d\_faststdio='undef';
> usefaststdio='define';

In perl build dir&#8203;:

\# cat config\.sh | grep faststdio
d\_faststdio='undef'
usefaststdio='undef'

but it seems this code to be still compiled\. Don't know whether
it is linked in the final product\.\.\.

> btw \- ICBW\.  If so\, im sure someone will correct me\.

What is ICBW?
\-\-
vda

**Comment by p5pRT** [comment]
### From @smpeters

> \[mjn3@&#8203;codepoet\.org \- Thu Oct 28 10&#8203;:45&#8203;:31 2004\]&#8203;:
> 
> On Thu\, Oct 28\, 2004 at 10&#8203;:42&#8203;:49AM \+0300\, Denis Vlasenko wrote&#8203;:
> > This part of perl source code caused build failure on uclibc\.
> > 
> > perl\-5\.8\.5/perlio\.c&#8203;:
> 
> See http&#8203;://www\.uclibc\.org/lists/uclibc/2004\-March/008512\.html
> 
> Manuel
> 

Rather than have everyone track it down\, here is the proposed patch from
that mailing list email\.

<details><summary>Inline Patch</summary>

```diff
--- perl-5.8.2/perlio.c-dist	2003-12-03 14:30:24.000000000 -0700
+++ perl-5.8.2/perlio.c	2003-12-03 14:32:10.000000000 -0700
@@ -2824,7 +2824,15 @@
     /* XXX this could use PerlIO_canset_fileno() and
      * PerlIO_set_fileno() support from Configure
      */
-#  if defined(__GLIBC__)
+#  if defined(__UCLIBC__)
+    /* uClibc must come before glibc because it defines __GLIBC__ as

```
</details>
well. */
\+\#  if defined\(\_\_MASK\_READING\)
\+    f\->\_\_filedes = \-1;
\+\#  else
\+    f\->filedes = \-1;
\+\#  endif
\+    return 1;
\+\#  elif defined\(\_\_GLIBC\_\_\)
&nbsp;     /\* There may be a better way for GLIBC&#8203;:
&nbsp;         \- libio\.h defines a flag to not close\(\) on cleanup
&nbsp;      \*/

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @smpeters

> \[stmpeters \- Thu Nov 04 08&#8203;:09&#8203;:26 2004\]&#8203;:
> 
> > \[mjn3@&#8203;codepoet\.org \- Thu Oct 28 10&#8203;:45&#8203;:31 2004\]&#8203;:
> > 
> > On Thu\, Oct 28\, 2004 at 10&#8203;:42&#8203;:49AM \+0300\, Denis Vlasenko wrote&#8203;:
> > > This part of perl source code caused build failure on uclibc\.
> > > 
> > > perl\-5\.8\.5/perlio\.c&#8203;:
> > 
> > See http&#8203;://www\.uclibc\.org/lists/uclibc/2004\-March/008512\.html
> > 
> > Manuel
> > 
> 
> Rather than have everyone track it down\, here is the proposed patch from
> that mailing list email\.
> 
> \-\-\- perl\-5\.8\.2/perlio\.c\-dist    2003\-12\-03 14&#8203;:30&#8203;:24\.000000000 \-0700
> \+\+\+ perl\-5\.8\.2/perlio\.c    2003\-12\-03 14&#8203;:32&#8203;:10\.000000000 \-0700
> @&#8203;@&#8203; \-2824\,7 \+2824\,15 @&#8203;@&#8203;
>      /\* XXX this could use PerlIO\_canset\_fileno\(\) and
>       \* PerlIO\_set\_fileno\(\) support from Configure
>       \*/
> \-\#  if defined\(\_\_GLIBC\_\_\)
> \+\#  if defined\(\_\_UCLIBC\_\_\)
> \+    /\* uClibc must come before glibc because it defines \_\_GLIBC\_\_ as
> well\. \*/
> \+\#  if defined\(\_\_MASK\_READING\)
> \+    f\->\_\_filedes = \-1;
> \+\#  else
> \+    f\->filedes = \-1;
> \+\#  endif
> \+    return 1;
> \+\#  elif defined\(\_\_GLIBC\_\_\)
>      /\* There may be a better way for GLIBC&#8203;:
>          \- libio\.h defines a flag to not close\(\) on cleanup
>       \*/    
> 


This was included with change \#23732\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
